lib/iris/_data_manager.py

```diff
@@ -254,6 +284,10 @@ def _realised_dtype_setter(self, realised_dtype):
                     emsg = ('Cannot set realised dtype, no lazy data '
                             'is available.')
                     raise ValueError(emsg)
+                if self._lazy_array.dtype.kind != 'f':
+                    emsg = ('Cannot set realised dtype for lazy data '
+                            'with {!r}.')
+                    raise ValueError(emsg.format(self._lazy_array.dtype))
```

Comment:
This check was missing, and was uncovered through testing the port of `fill_value` from the `Cube` to the `DataManager`